File: mailblastr-php/src/Resources/Contacts.php
<?php

declare(strict_types=1);

namespace Mailblastr\Resources;

use Mailblastr\Client;

/**
 * Contacts are DOMAIN-FIRST: each sending domain has its own contact pool (the
 * same address on two domains is two records with separate consent).
 *
 * - Pass 'domain' to use the flat `/contacts` API (REQUIRED there on
 *   create/list; on get/update/remove it disambiguates an EMAIL id across pools).
 * - Pass 'audienceId' instead to target a specific audience via the nested
 *   `/audiences/:id/contacts` API.
 */
class Contacts extends Resource
{
    /**
     * Create a contact. POST /contacts (flat, 'domain' required) or
     * POST /audiences/:id/contacts (when 'audienceId' is given).
     * Returns the slim ack ['object' => 'contact', 'id' => …].
     *
     * @param array $params domain|audienceId + email (required), first_name,
     *                      last_name, unsubscribed, properties.
     */
    public function create(array $params): array
    {
        $audienceId = $params['audienceId'] ?? null;
        unset($params['audienceId']);
        if ($audienceId !== null) {
            // The nested audience route derives its pool from the path; only
            // the flat route takes `domain` in the body.
            unset($params['domain']);
            return $this->client->request('POST', '/audiences/' . Client::e((string) $audienceId) . '/contacts', $params);
        }
        return $this->client->request('POST', '/contacts', $params);
    }

    /**
     * Retrieve a contact by id or email. An id is exact; an EMAIL can exist in
     * several domains' pools, so pass 'domain' to pick the pool (omitted → the
     * oldest match anywhere).
     *
     * @param array $params ['id' => contact id or email, 'domain' => …, 'audienceId' => …]
     */
    public function get(array $params): array
    {
        $id = Client::e((string) ($params['id'] ?? ''));
        if (!empty($params['audienceId'])) {
            return $this->client->request(
                'GET',
                '/audiences/' . Client::e((string) $params['audienceId']) . '/contacts/' . $id
            );
        }
        $qs = isset($params['domain']) ? '?domain=' . rawurlencode((string) $params['domain']) : '';
        return $this->client->request('GET', '/contacts/' . $id . $qs);
    }

    /**
     * List contacts. Domain-first: 'domain' is required on the flat `/contacts`
     * list (names the pool); pass 'audienceId' for the nested list instead.
     * 'segment_id' restricts to members of that segment (both APIs).
     *
     * @param array $params domain|audienceId, limit, after, before, segment_id.
     */
    public function list(array $params = []): array
    {
        $audienceId = $params['audienceId'] ?? null;
        $query = [];
        if ($audienceId === null && isset($params['domain'])) {
            $query['domain'] = $params['domain'];
        }
        foreach (['limit', 'after', 'before', 'segment_id'] as $key) {
            if (isset($params[$key])) {
                $query[$key] = $params[$key];
            }
        }
        $base = $audienceId !== null
            ? '/audiences/' . Client::e((string) $audienceId) . '/contacts'
            : '/contacts';
        return $this->client->request('GET', $base . $this->client->query($query));
    }

    /**
     * Bulk-import contacts from an array. Upserts by email; max 10,000 per
     * call. Domain-first, like create(): pass 'domain' for the flat
     * POST /contacts/batch, or 'audienceId' for
     * POST /audiences/:id/contacts/batch.
     *
     * Prefer this over a create() loop for many contacts: one batch takes the
     * account's contact-limit lock once, a loop takes it per contact.
     *
     * @param array $params ['domain' => … | 'audienceId' => …,
     *                      'contacts' => [ … ],
     *                      'on_conflict' => 'upsert'|'skip' ('skip' leaves
     *                      existing contacts untouched; default 'upsert')]
     */
    public function batch(array $params): array
    {
        $qs = isset($params['on_conflict'])
            ? '?on_conflict=' . rawurlencode((string) $params['on_conflict'])
            : '';
        $body = ['contacts' => $params['contacts'] ?? []];
        if (isset($params['audienceId'])) {
            return $this->client->request(
                'POST',
                '/audiences/' . Client::e((string) $params['audienceId']) . '/contacts/batch' . $qs,
                $body
            );
        }
        // The nested route derives its pool from the path; only the flat route
        // takes 'domain' (in the body, same as POST /contacts).
        if (isset($params['domain'])) {
            $body['domain'] = $params['domain'];
        }
        return $this->client->request('POST', '/contacts/batch' . $qs, $body);
    }

    /**
     * Bulk-import contacts from CSV. Upserts by email. By default every
     * non-builtin CSV column is auto-registered as a custom property (so
     * `company`, `plan`, … survive and become {{merge}} tags); pass
     * 'create_properties' => false for strict mode.
     * POST /audiences/:id/contacts/import
     *
     * Two input modes: inline 'csv' text (max 5 MB / 10,000 rows), or
     * 'storage_key' from {@see createImportUpload()} for a file uploaded straight to
     * storage (no size limit beyond the upload cap; rows beyond the account's
     * remaining contact capacity are reported as limit_skipped instead of
     * failing).
     *
     * @param array $params ['audienceId' => …,
     *                      'csv' => … OR 'storage_key' => …,
     *                      'file_name' => name recorded for the inline CSV,
     *                      'on_conflict' => 'upsert'|'skip',
     *                      'create_properties' => bool,
     *                      'segment_id' => add every imported email to this
     *                      segment (must belong to the same audience)]
     */
    public function import(array $params): array
    {
        $query = [];
        if (isset($params['on_conflict'])) {
            $query['on_conflict'] = $params['on_conflict'];
        }
        if (($params['create_properties'] ?? null) === false) {
            $query['create_properties'] = 'false';
        }
        if (isset($params['segment_id'])) {
            $query['segment_id'] = $params['segment_id'];
        }

        $body = [];
        foreach (['csv', 'storage_key', 'file_name'] as $key) {
            if (isset($params[$key])) {
                $body[$key] = $params[$key];
            }
        }
        if ($body === []) {
            $body['csv'] = '';
        }

        return $this->client->request(
            'POST',
            '/audiences/' . Client::e((string) $params['audienceId']) . '/contacts/import' . $this->client->query($query),
            $body
        );
    }

    /**
     * Mint a presigned URL for uploading a CSV straight to storage, so a large
     * file never passes through the API. PUT the file to the returned
     * 'upload_url', then hand the returned 'storage_key' to {@see import()}.
     * POST /audiences/:id/contacts/import/upload
     *
     * The returned 'upload_url' is a bearer credential — do not log it.
     *
     * @param array $params ['audienceId' => …, 'filename' => must end in .csv,
     *                      'size' => file size in bytes (max 256 MB)]
     */
    public function createImportUpload(array $params): array
    {
        $body = [];
        foreach (['filename', 'size'] as $key) {
            if (isset($params[$key])) {
                $body[$key] = $params[$key];
            }
        }
        return $this->client->request(
            'POST',
            '/audiences/' . Client::e((string) $params['audienceId']) . '/contacts/import/upload',
            $body
        );
    }

    /**
     * Update a contact. Returns the slim ack ['object' => 'contact', 'id' => …].
     * On the flat API, pass 'domain' when 'id' is an EMAIL (disambiguates
     * across pools).
     *
     * @param array $params ['id' => contact id or email, 'audienceId' => …,
     *                      'domain' => …] + first_name, last_name,
     *                      unsubscribed, properties.
     */
    public function update(array $params): array
    {
        $id = (string) ($params['id'] ?? '');
        $audienceId = $params['audienceId'] ?? null;
        unset($params['id'], $params['audienceId']);
        if ($audienceId !== null) {
            // The nested audience route derives its pool from the path; the
            // flat route takes an optional `domain` in the body.
            unset($params['domain']);
            return $this->client->request(
                'PATCH',
                '/audiences/' . Client::e((string) $audienceId) . '/contacts/' . Client::e($id),
                $params
            );
        }
        return $this->client->request('PATCH', '/contacts/' . Client::e($id), $params);
    }

    /**
     * Delete a contact. On the flat API, pass 'domain' when 'id' is an EMAIL
     * (disambiguates across pools).
     *
     * @param array $params ['id' => contact id or email, 'domain' => …, 'audienceId' => …]
     */
    public function remove(array $params): array
    {
        $id = Client::e((string) ($params['id'] ?? ''));
        if (!empty($params['audienceId'])) {
            return $this->client->request(
                'DELETE',
                '/audiences/' . Client::e((string) $params['audienceId']) . '/contacts/' . $id
            );
        }
        $qs = isset($params['domain']) ? '?domain=' . rawurlencode((string) $params['domain']) : '';
        return $this->client->request('DELETE', '/contacts/' . $id . $qs);
    }

    /** Add a contact to a segment. POST /contacts/:id/segments/:segmentId */
    public function addToSegment(string $id, string $segmentId): array
    {
        return $this->client->request(
            'POST',
            '/contacts/' . Client::e($id) . '/segments/' . Client::e($segmentId)
        );
    }

    /** Remove a contact from a segment. DELETE /contacts/:id/segments/:segmentId */
    public function removeFromSegment(string $id, string $segmentId): array
    {
        return $this->client->request(
            'DELETE',
            '/contacts/' . Client::e($id) . '/segments/' . Client::e($segmentId)
        );
    }

    /**
     * List the segments a contact belongs to. GET /contacts/:id/segments
     *
     * @param array $params Optional cursor pagination (limit, after, before).
     *                      With no params the list comes back in ONE page
     *                      instead of a 20-row one, still capped at 1000 rows
     *                      with 'has_more' reporting the truncation.
     */
    public function listSegments(string $id, array $params = []): array
    {
        return $this->client->request(
            'GET',
            '/contacts/' . Client::e($id) . '/segments' . $this->paginationQuery($params)
        );
    }

    /**
     * Get a contact's topic subscriptions. GET /contacts/:id/topics
     *
     * @param array $params Optional cursor pagination (limit, after, before).
     *                      With no params the list comes back in ONE page
     *                      instead of a 20-row one, still capped at 1000 rows
     *                      with 'has_more' reporting the truncation.
     */
    public function getTopics(string $id, array $params = []): array
    {
        return $this->client->request(
            'GET',
            '/contacts/' . Client::e($id) . '/topics' . $this->paginationQuery($params)
        );
    }

    /**
     * Update a contact's topic subscriptions. PATCH /contacts/:id/topics
     *
     * @param array $payload ['topics' => [['id' => …, 'subscription' => 'opt_in'|'opt_out'], …]]
     */
    public function updateTopics(string $id, array $payload): array
    {
        return $this->client->request('PATCH', '/contacts/' . Client::e($id) . '/topics', $payload);
    }
}

File: mailblastr-ruby/lib/mailblastr/contacts.rb
# frozen_string_literal: true

module Mailblastr
  # Contacts are DOMAIN-FIRST: each sending domain has its own contact pool,
  # so the flat /contacts API takes `domain` (required on create/list). The
  # nested audience variants (`audience_id:`) derive the pool from the path.
  module Contacts
    class << self
      # Create a contact. POST /contacts (flat, `domain` required) or
      # POST /audiences/:id/contacts when `audience_id` is given.
      #   Mailblastr::Contacts.create({ domain: "yourdomain.com", email: "a@b.com" })
      #   Mailblastr::Contacts.create({ audience_id: "aud_1", email: "a@b.com" })
      def create(params)
        audience_id = Client.opt(params, :audience_id)
        if audience_id
          body = Client.without(params, :audience_id, :domain)
          Client.request(:post, "/audiences/#{Client.path_escape(audience_id)}/contacts", body: body)
        else
          Client.require_domain!(params, "Contacts.create (flat /contacts API)")
          Client.request(:post, "/contacts", body: Client.without(params, :audience_id))
        end
      end

      # Retrieve a contact by id (exact) or by email. An email can exist in
      # several domains' pools, so pass `domain` to pick the pool.
      #   Contacts.get({ id: "cont_1" })
      #   Contacts.get({ id: "a@b.com", domain: "yourdomain.com" })
      #   Contacts.get({ id: "cont_1", audience_id: "aud_1" })
      def get(params)
        id = Client.path_escape(Client.opt(params, :id))
        audience_id = Client.opt(params, :audience_id)
        return Client.request(:get, "/audiences/#{Client.path_escape(audience_id)}/contacts/#{id}") if audience_id

        domain = Client.opt(params, :domain)
        query = domain ? { domain: domain } : nil
        Client.request(:get, "/contacts/#{id}", query: query)
      end

      # List contacts. Flat /contacts requires `domain` (names the pool);
      # pass `audience_id` instead to use the nested API. `segment_id`
      # filters either variant; limit/after/before paginate.
      def list(params = {})
        audience_id = Client.opt(params, :audience_id)
        query = Client.pagination(params)
        segment_id = Client.opt(params, :segment_id)
        query[:segment_id] = segment_id if segment_id

        if audience_id
          Client.request(:get, "/audiences/#{Client.path_escape(audience_id)}/contacts", query: query)
        else
          query = { domain: Client.require_domain!(params, "Contacts.list (flat /contacts API)") }.merge(query)
          Client.request(:get, "/contacts", query: query)
        end
      end

      # Update a contact (id or email). PATCH /contacts/:id, or the nested
      # route when `audience_id` is given. On the flat API pass `domain` when
      # `id` is an email (disambiguates across pools).
      #   Contacts.update({ id: "cont_1", unsubscribed: true })
      def update(params)
        id = Client.path_escape(Client.opt(params, :id))
        audience_id = Client.opt(params, :audience_id)
        if audience_id
          body = Client.without(params, :audience_id, :domain, :id)
          Client.request(:patch, "/audiences/#{Client.path_escape(audience_id)}/contacts/#{id}", body: body)
        else
          Client.request(:patch, "/contacts/#{id}", body: Client.without(params, :audience_id, :id))
        end
      end

      # Delete a contact. DELETE /contacts/:id (pass `domain` when `id` is an
      # email), or the nested route when `audience_id` is given.
      def delete(params)
        id = Client.path_escape(Client.opt(params, :id))
        audience_id = Client.opt(params, :audience_id)
        return Client.request(:delete, "/audiences/#{Client.path_escape(audience_id)}/contacts/#{id}") if audience_id

        domain = Client.opt(params, :domain)
        query = domain ? { domain: domain } : nil
        Client.request(:delete, "/contacts/#{id}", query: query)
      end

      # Bulk-import contacts from an array (upsert by email; max 10,000).
      #
      # Domain-first, like .create: pass :domain for the flat
      # POST /contacts/batch, or :audience_id for POST /audiences/:id/contacts/batch.
      # Prefer this over a .create loop for many contacts — one batch takes the
      # account's contact-limit lock once, a loop takes it per contact.
      #   Contacts.batch({ domain: "yourdomain.com", contacts: [{ email: "a@b.com" }] })
      #   Contacts.batch({ audience_id: "aud_1", contacts: [{ email: "a@b.com" }], on_conflict: "skip" })
      def batch(params)
        audience_id = Client.opt(params, :audience_id)
        # "" is truthy in Ruby but names no audience — treat it as absent and
        # take the flat route, matching the other MailBlastr SDKs.
        audience_id = nil if audience_id == ""
        query = {}
        on_conflict = Client.opt(params, :on_conflict)
        query[:on_conflict] = on_conflict if on_conflict
        body = { contacts: Client.opt(params, :contacts) }
        return Client.request(
          :post,
          "/audiences/#{Client.path_escape(audience_id)}/contacts/batch",
          body: body,
          query: query
        ) if audience_id

        # The nested route derives its pool from the path; only the flat route
        # takes :domain (in the body, same as POST /contacts).
        domain = Client.opt(params, :domain)
        body[:domain] = domain unless domain.nil?
        Client.request(:post, "/contacts/batch", body: body, query: query)
      end

      # Bulk-import contacts from CSV (header row optional; upsert by email).
      # Non-builtin columns auto-register as custom properties unless
      # `create_properties: false`. Pass `segment_id` to also add every
      # imported email to one of this audience's segments.
      # POST /audiences/:id/contacts/import
      #
      # Inline CSV text (capped at 5 MB and 10,000 rows):
      #   Contacts.import({ audience_id: "aud_1", csv: "email\na@b.com" })
      # Or a file already uploaded via create_import_upload (no row cap — the
      # overflow past your contact limit comes back as `limit_skipped`):
      #   Contacts.import({ audience_id: "aud_1", storage_key: key })
      def import(params)
        audience_id = Client.opt(params, :audience_id)
        query = Client.filters(params, :on_conflict, :segment_id)
        query[:create_properties] = "false" if Client.opt(params, :create_properties) == false
        body = Client.filters(params, :csv, :file_name, :storage_key)
        Client.request(
          :post,
          "/audiences/#{Client.path_escape(audience_id)}/contacts/import",
          body: body,
          query: query
        )
      end

      # Mint a presigned direct-upload URL for a CSV too large to inline
      # (up to 256 MB). Upload the file to `upload_url`, then pass the returned
      # `storage_key` to Contacts.import.
      # POST /audiences/:id/contacts/import/upload — params: { filename:, size: }
      # The `upload_url` is a bearer credential — do not log it.
      def create_import_upload(params)
        audience_id = Client.opt(params, :audience_id)
        Client.request(
          :post,
          "/audiences/#{Client.path_escape(audience_id)}/contacts/import/upload",
          body: Client.without(params, :audience_id)
        )
      end

      # Add a contact to a segment. POST /contacts/:id/segments/:segment_id
      def add_to_segment(contact_id, segment_id)
        Client.request(:post, "/contacts/#{Client.path_escape(contact_id)}/segments/#{Client.path_escape(segment_id)}")
      end

      # Remove a contact from a segment. DELETE /contacts/:id/segments/:segment_id
      def remove_from_segment(contact_id, segment_id)
        Client.request(:delete, "/contacts/#{Client.path_escape(contact_id)}/segments/#{Client.path_escape(segment_id)}")
      end

      # List the segments a contact belongs to — items carry id/name/created_at
      # only, not the full segment object. GET /contacts/:id/segments
      def list_segments(contact_id, params = {})
        Client.request(
          :get,
          "/contacts/#{Client.path_escape(contact_id)}/segments",
          query: Client.pagination(params)
        )
      end

      # Get a contact's topic subscriptions. GET /contacts/:id/topics
      def get_topics(contact_id, params = {})
        Client.request(
          :get,
          "/contacts/#{Client.path_escape(contact_id)}/topics",
          query: Client.pagination(params)
        )
      end

      # Update a contact's topic subscriptions. PATCH /contacts/:id/topics
      #   Contacts.update_topics("cont_1", { topics: [{ id: "top_1", subscription: "opt_in" }] })
      def update_topics(contact_id, params)
        Client.request(:patch, "/contacts/#{Client.path_escape(contact_id)}/topics", body: params)
      end
    end
  end
end
